kep/640-error-pods-rescheduling/README.md

```diff
@@ -0,0 +1,166 @@
+# Error pods rescheduling
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+- [Design Details](#design-details)
+  - [Rescheduler-controller](#rescheduler-controller)
+    - [Scheduled nodes list and current rescheduling times definition](#scheduled-nodes-list-and-current-rescheduling-times-definition)
+  - [Podrescheduling](#podrescheduling)
+  - [A concrete use case](#a-concrete-use-case)
+  - [Notes](#notes)
+- [Known limitations](#known-limitations)
+- [Test Plan](#test-plan)
+- [Graduation Criteria](#graduation-criteria)
+  - [Alpha](#alpha)
+  - [Beta](#beta)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+## Summary
+
+This proposal outlines a solution for rescheduling pods that have encountered errors, such as pods being OOM-killed or experiencing errors related to NUMA or devices.
+
+## Motivation
+
+The current Kubernetes scheduler performs scheduling for a pod only once. Consequently, if a scheduled pod encounters errors during booting or while running, the scheduler won't intervene to reschedule it to other nodes. 
+The in-place restart strategy provided by kubelet only handles local restarts, which often fall short of resolving issues. 
+For instance, a pod might get OOM killed due to insufficient host-node memory, or it could be rejected by kubelet during startup because of mismatches in NUMA or other device requirements. 
+In such cases, local restarts are ineffective. Manual intervention becomes necessary, leading to increased operational complexity.
+
+This situation contradicts Kubernetes' claim of automatic fault tolerance capabilities, diminishing the reliability and high availability of the service. 
+Therefore, we are motivated to develop rescheduling plugins based on the proposed scheduler framework. These plugins can reschedule pods that encounter errors to different nodes, excluding the nodes where they were previously scheduled.
+
+### Goals
+
+* Reschedule already scheduled pods (encounter some errors) to other nodes automatically.
+* Implement pod rescheduling plugins in PreFilter extension point to prefilter nodes that the pod has already scheduled.
+* Pod rescheduling plugin in this proposal has no conflict with default scheduler's plugins.
+
+### Non-Goals
+
+* Reschedule pods controlled by daemonset or other workload resources which don't exist in default k8s controller.
```

Comment:
is this to avoid changes to existing controllers? (sorry if asked previously, if so just link the answer please)

Reply:
We don't need to change existing controllers. We reschedule thoses pods which controlled by existing k8s default controllers like deployment or statefulset, but don't reschedule pods controlled by user's newly defined controllers.

As for the daemonset controller, it ensures that all nodes run a copy of a pod always, so we don't reschedule thoses pods controlled by damoneset.

Reply:
OK, so this proposal seems a subset of the kubelet-triggers-reschedule one which we're working on. Avoiding any change to kubernetes core is a pretty severe limitation.
In general I think it would be good to start with this approach in the scheduler-plugins and see how it performs. 
We can later generalize (or try to) the solution incorporating the lessons learned here. And if things go well, maybe have something in kubernetes core someday in the future.

Reply:
Ok, thanks for your reply. As you mentioned above, [here](https://docs.google.com/document/d/1-wJhiNy84w7tzFdo9HqwTu5DrVSuXFLGTUv8FBiRAAc/edit#heading=h.iyqdsaanqjtg), i have a great interest in cooperating in this proposal.
> Mostly general comments about the design and the possible interactions with a very similar work recently started from DRA conversations and linked multiple times.
> 
> Would you be interested in cooperating in such effort? Full disclosure: is not yet clear if the _proposal_ can be sent in time for 1.30 because EU/NA holidays. I'm aiming to be ready in time for 1.30 but I can't make promises. This effort aims to change core k8s so it can take a while.
> 
> I'll have a deeper look at the implementation ASAP.

So how do I get involved in this project? Thank you.

Reply:
Hi! for starters, please join the #wg-batch channel on k8s slack and perhaps join the conversations here (I'll be updating regularly but please be aware of the upcoming EU/NA holidays and limited availability).
I'll open the aforementioned design doc for comments as soon as the remaining gaps are open.

Reply:
Ok, thanks for your reply. I have joined the #wg-batch channel on k8s slack.
I have updated the latest KEP version, is there anything else I need to change about this KEP?
Thanks a lot for your review!

Reply:
I will have another pass as soon as I can, but generally speaking I don't have any major objection.

Please note however that while I for myself I'm ok with the approach outlined here, it seems to me that basically this proposal is adding a new controller for pods which kinda overlaps with deployment controller - it manages pod retries after all, reminiscent of how a deployment would do. Or, in general, there are similarities. For this reason I think more review, especially from more senior contributors of the project is required.

Reply:
Ok, thanks for your reply. The Rescheduler-controller will not be pushed to this project, but only podrescheduling scheduler-plugins be pushed.